packages/dev/core/src/Rendering/IBLShadows/iblShadowsVoxelTracingPass.ts

```diff
@@ -254,76 +261,98 @@ export class _IblShadowsVoxelTracingPass {
             defines += "#define VOXEL_MARCH_DIAGNOSTIC_INFO_OPTION 1u\n";
         }
         const isWebGPU = this._engine.isWebGPU;
-        const ppOptions: PostProcessOptions = {
-            width: this._engine.getRenderWidth(),
-            height: this._engine.getRenderHeight(),
-            textureFormat: Constants.TEXTUREFORMAT_RG,
-            textureType: Constants.TEXTURETYPE_UNSIGNED_BYTE,
+        const textureOptions: IProceduralTextureCreationOptions = {
+            type: Constants.TEXTURETYPE_UNSIGNED_BYTE,
+            format: Constants.TEXTUREFORMAT_RG,
             samplingMode: Constants.TEXTURE_NEAREST_SAMPLINGMODE,
-            uniforms: ["viewMtx", "projMtx", "invProjMtx", "invViewMtx", "wsNormalizationMtx", "shadowParameters", "offsetDataParameters", "sssParameters", "shadowOpacity"],
-            samplers: ["voxelGridSampler", "icdfySampler", "icdfxSampler", "blueNoiseSampler", "worldNormalSampler", "depthSampler", "worldPositionSampler"],
-            defines: defines,
-            engine: this._engine,
-            reusable: false,
+            generateDepthBuffer: false,
             shaderLanguage: isWebGPU ? ShaderLanguage.WGSL : ShaderLanguage.GLSL,
-            extraInitializations: (useWebGPU: boolean, list: Promise<any>[]) => {
-                if (useWebGPU) {
-                    list.push(import("../../ShadersWGSL/iblShadowVoxelTracing.fragment"));
+            extraInitializationsAsync: async () => {
+                if (isWebGPU) {
+                    await Promise.all([import("../../ShadersWGSL/iblShadowVoxelTracing.fragment")]);
                 } else {
-                    list.push(import("../../Shaders/iblShadowVoxelTracing.fragment"));
+                    await Promise.all([import("../../Shaders/iblShadowVoxelTracing.fragment")]);
                 }
             },
         };
-        this._outputPP = new PostProcess("voxelTracingPass", "iblShadowVoxelTracing", ppOptions);
-        this._outputPP.autoClear = false;
-        this._outputPP.onApplyObservable.add((effect) => {
-            this._updatePostProcess(effect, this._scene.activeCamera!);
+        this._outputTexture = new ProceduralTexture(
+            "voxelTracingPass",
+            {
+                width: this._engine.getRenderWidth(),
+                height: this._engine.getRenderHeight(),
+            },
+            "iblShadowVoxelTracing",
+            this._scene,
+            textureOptions,
+            false,
+            false,
+            Constants.TEXTURETYPE_UNSIGNED_INT
+        );
+        this._outputTexture.refreshRate = -1;
+        this._outputTexture.autoClear = false;
+        this._outputTexture.defines = defines;
+        // Need to set all the textures first so that the effect gets created with the proper uniforms.
+        this._update(this._scene.activeCamera!);
+
+        this._scene.onBeforeCameraRenderObservable.add(() => {

```

Comment:
I don't really like this bit of logic but I couldn't figure out a way to get the PT to render when I want. I want it to render after the geometry buffer is finished but before the main scene. The `onAfterRenderTargetsRenderObservable` seems to be fired after the customRenderTargets on the scene as well as after the geometry buffer renders. Hence, my logic of listening to `onBeforeCameraRenderObservable` and then binding to one occurrence of  `onAfterRenderTargetsRenderObservable`.
Is there a better way of doing this?

Reply:
The current code wouldn't work if there were multiple cameras in the scene, because `onBeforeCameraRenderObservable` is called for each camera. Since `onAfterRenderTargetsRenderObservable` is called after the scene-level RTs have been processed, and also after the camera-level RTs have been processed, you could do something like this:

```typescript
let counter = 0;

scene.onBeforeRenderObservable.add(() => {
    counter = 0;
});
scene.onAfterRenderTargetsRenderObservable.add(() => {
    if (++counter == 2) {
        <your code>
    }
});
```
It's not ideal, but without the render graph, I can't think of a cleaner way to do it...

Let us know when the PR is ready for review!

Reply:
Thank you for the suggestion. I've implemented this now.